File: B_Bessie_and_MEX.py
def find_mex(hashmap,mexx):
    while mexx in hashmap:
        mexx+=1
    return mexx
    
def construct_permutation(n, a):
    p = []
    mex=0
    hashmap = {}
    for i in range(n):
        mex = find_mex(p,mex)
        if( mex - a[i] < 0 or mex - a[i] in p):

            p.append(mex)
            hashmap[mex] = True
            mex = find_mex(hashmap,mex)

        else:
            p.append(mex - a[i])
            hashmap[mex - a[i]] = True
    return p

# Read input
t = int(input())
for _ in range(t):
    n = int(input())
    a = list(map(int, input().split()))
    
    # Construct permutation
    result = construct_permutation(n, a)
    
    # Print the permutation
    print(*result)

File: B_Bessie_and_MEX.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

int find_mex(std::unordered_map<int, bool>& hashmap, int mexx) {
    while (hashmap.find(mexx) != hashmap.end()) {
        mexx++;
    }
    return mexx;
}

std::vector<int> construct_permutation(int n, std::vector<int>& a) {
    std::vector<int> p;
    int mex = 0;
    std::unordered_map<int, bool> hashmap;
    for (int i = 0; i < n; ++i) {
        mex = find_mex(hashmap, mex);
        if (mex - a[i] < 0 || hashmap.find(mex - a[i]) != hashmap.end()) {
            p.push_back(mex);
            hashmap[mex] = true;
            mex = find_mex(hashmap, mex);
        } else {
            p.push_back(mex - a[i]);
            hashmap[mex - a[i]] = true;
        }
    }
    return p;
}

int main() {
    int t;
    std::cin >> t;
    while (t--) {
        int n;
        std::cin >> n;
        std::vector<int> a(n);
        for (int i = 0; i < n; ++i) {
            std::cin >> a[i];
        }
        
        // Construct permutation
        std::vector<int> result = construct_permutation(n, a);
        
        // Print the permutation
        for (int i = 0; i < n; ++i) {
            std::cout << result[i] << " ";
        }
        std::cout << std::endl;
    }
    return 0;
}
